Create more detailed OSR documentation

gita-omr opened the pull request:
- describe all OSR-based optimizations using concepts in the OMR document

gita-omr commented:
Added commit message and removed line ending spaces.

gita-omr commented:
Fixed line endings.

gita-omr commented:
Rebased.

hzongaro reviewed (commented):
Started reviewing, but then I had to put it aside.  I thought I would submit my comments thus far, so they don't get lost.

gita-omr commented on doc/compiler/hcr/NextGenHCR.md:
Sure!

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+
+# FSD
+
+Previously, OpenJ9  only allowed debugging optimized code via the Full-Speed Debug (FSD) option.
```
This whole paragraph is from the old document. To be honest, I don't recall when exactly OSR based FSD was implemented.  Maybe I can  say "Originally" ?

gita-omr replied:
Similar, re the FSD option, it's what the old doc said. Need to find out more about it...

gita-omr replied:
Changed to:
_Until around 2014,_ OpenJ9 only allowed debugging optimized code in the _old_ Full-Speed Debug (FSD) mode.
In _the old_ FSD mode...

_New_ FSD is implemented using involuntary OSR

jdmpapin replied:
That clarifies it for me, but I'll leave it to @hzongaro to decide whether this comment is resolved

hzongaro replied:
Sounds good.

gita-omr commented:
Thanks @hzongaro for the review. I will address it soon. @jdmpapin could you please review as well?

jdmpapin commented on doc/compiler/osr/OSR.md:
After this we'll also do OSR exception edge removal, which removes all OSR exception edges except those that originate from OSR induce blocks

I think it would be good to show this example CFG after that transformation as well. In particular, the OSR catch block and OSR code block for B will become unreachable and get removed

gita-omr replied:
Right. Just got tired of drawing :) Will add another CFG.

gita-omr replied:
done. Please take a look.

jdmpapin commented on doc/compiler/osr/OSR.md:
I'm not sure what you mean by *performing as usual*

gita-omr replied:
This is an old text :) Let me see how it can be improved.

gita-omr replied:
This is from the old doc. I think the next sentence explains: "This is a debug mode to evaluate OSR overhead and it will not correctly implement redefinitions." I guess "usual" means "as described above". I will replace.

jdmpapin replied:
Oh, I understand. In that case, what about *working* as described above? In this context *performing* sounds to me like it's about the performance of the generated code

jdmpapin commented on doc/compiler/osr/OSR.md:
*pre-execution* OSR (if you agree with my corresponding comment on the OMR documentation)

gita-omr replied:
done

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+Previously, OpenJ9  only allowed debugging optimized code via the Full-Speed Debug (FSD) option.
+In FSD mode, a few simple optimizations such as trivial inlining, basic block extension, and local common sub-expression elimination were enabled.
+The implementation faithfully maintained the correct stack slot values as well as the bytecode index ordering as the interpreter would have
+at any de-compilation point. For instance, inlining was only done if there is no possible GC point in the code being inlined.
```
What about "at any point where the method need to transition to interpreter" ? @hzongaro

hzongaro replied:
I think that would still need to distinguish between situations where JIT-compiled code is calling an interpreted method from situations where we transition from running the compiled version of the method to running the interpreted version of the method while the method is running.

hzongaro replied:
I'm not going to let this comment hold up the change.  Marking it as resolved.

gita-omr commented on doc/compiler/osr/OSR.md:
Sounds good to me. I will use it. Thanks a lot!

gita-omr replied:
done

gita-omr commented on doc/compiler/osr/OSR.md:
Changed here:

## OSR and virtual guard merging

As an optimization, we can mark some existing virtual guard as an OSR guard, therefore merging them together. This is achieved by making virtual guard kill fear. _In this case, the taken side of a merged OSR guard will be a cold call (vs. OSR induce block)_. The disadvantage ...

gita-omr commented on doc/compiler/osr/OSR.md:
Changing to: 
"_As an optimization, we might have a store of the call result into an auto other than a pending push temp. This is because_ call results are often immediately stored to autos in the bytecode."

gita-omr commented on doc/compiler/osr/OSR.md:
I thought the text you suggested would be sufficient. It's definitely would be much more complex to provide individual tracking but I am not sure I can think of all the implementation details right now.

gita-omr commented on doc/compiler/osr/OSR.md:
Oh, I mentioned it because we specifically introduced a change where we don't do guard merging for Vector API methods since it resulted in extra virtual guards that could've been eliminated otherwise. Maybe we don't need that change anymore?

jdmpapin replied:
Ah, no, that change is still needed because the OSR guard would still remain, and it would still target the cold call. I suppose it might possibly be removed later based on OSR guard analysis, but we can't rely on the OSR guard being redundant whenever the virtual guard is guaranteed to pass.

Maybe say that the merged guard cannot be *completely* removed anymore? But even that seems misleading given the possibility that the OSR guard could happen to be redundant. Taking a stab at it:
> The disadvantage of guard merging is that a later type analysis that would be sufficient to remove a virtual guard won't be able to also remove the corresponding OSR guard, which likely prevents the elimination of a control flow diamond.

gita-omr commented on doc/compiler/osr/OSR.md:
It might be due the confusing formatting. This sub-section is under the nextGenHCR section and it's mentioned at the beginning that it's a voluntary OSR. But I am not sure how to make this formatting more clear.

gita-omr replied:
except for relying on the header's fond size.

gita-omr replied:
I will check the helpers...

gita-omr replied:
Changed to:
"## OSR induce blocks

The main purpose of the OSR induce blocks is to voluntarily induce OSR.
In addition, since OSR catch blocks apply for any possible transition points in a method, the OSR induce block allow bookkeeping to be specified at an individual OSR point.
It contains **irrelevant stores**, which store default values to symrefs.
These ensure that variables that are dead at a specific OSR point don't spuriously appear to be live based on uses in the OSR code block.
Moreover, the block potentially contains rematerializaion of live symrefs and monitor enters, to further reduce the constraints on other optimizations.
The OSR induce blocks end with an OSR induce (a call to jitInduceOSRAtCurrentPC helper) which transitions to the VM, and an exception edge to the OSR catch block.
Once the induce helper is called, we transition to the VM."

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+
+# Introduction
+
+This document describes several OpenJ9 optimizations that use OSR mechanism described in [OMR OSR document](https://github.com/eclipse/omr/blob/master/doc/compiler/osr/OSR.md). These optimizations include:
```
done

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+
+FSD is implemented using involuntary OSR. The set of OSR yield points includes all GC points: call nodes, nodes creating an object instance,
+various checks including NULL checks and async checks, any debugging hooks such as method entries and exits, etc.
+When debugging event occurs, next yield to VM at a GC point that corresponds to that event causes an exception and VM transfers control
```
fixed

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+
+FSD is implemented using involuntary OSR. The set of OSR yield points includes all GC points: call nodes, nodes creating an object instance,
+various checks including NULL checks and async checks, any debugging hooks such as method entries and exits, etc.
+When debugging event occurs, next yield to VM at a GC point that corresponds to that event causes an exception and VM transfers control
```
fixed

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+FSD is implemented using involuntary OSR. The set of OSR yield points includes all GC points: call nodes, nodes creating an object instance,
+various checks including NULL checks and async checks, any debugging hooks such as method entries and exits, etc.
+When debugging event occurs, next yield to VM at a GC point that corresponds to that event causes an exception and VM transfers control
+to the correct catch block. For example. if "break on throw" is set, next check node that throws an exception will cause control to be switch to the
```
fixed

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+Under HCR, agents may redefine the implementation of a Java method in terms of bytecodes, but may not alter the inheritance hierarchy or shape of classes
+(no adding or removing fields or methods for example). If the code of a method is changed while we are executing that method, we are allowed to finish executing the old code.
+For example, assume that A calls B and B is inlined into A. If B's definition changes while we are executing B's code (inlined in A), we can finish it and
+the next call to B (from A or otherwise) need to call the new definition of B.
```
fixed

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+
+The traditional means of implementing HCR support in the JIT was to add an HCR guard to every inlined method.
+The guard would patch if the method it protected was redefined - this stops us from executing the inlined method body and instead calls the VM to run the new method implementation.
+The control-flow diamond this creates is the reason from the throughput performance overhead mentioned previously -
```
fixed

gita-omr commented on doc/compiler/osr/OSR.md:
fixed

gita-omr commented on doc/compiler/osr/OSR.md:
fixed

gita-omr commented on doc/compiler/osr/OSR.md:
done

gita-omr commented on doc/compiler/osr/OSR.md:
done

gita-omr commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,237 @@
+
+* Full Speed Debug (FSD)
+* Next Generation Hot Code Replacement (nextGenHCR)
+* Final Static Field Folding
```
done

gita-omr commented on doc/compiler/osr/OSR.md:
done

gita-omr commented on doc/compiler/osr/OSR.md:
done

gita-omr commented on doc/compiler/osr/OSR.md:
done

jdmpapin replied:
This now says:
> The fall through path will be a path to a fear point without going through any other OSR guards

which still doesn't seem right to me. It's just the start of some such path. I think the wording now suggests that on fallthrough we *will* go a fear point, but in some cases we only *might*. How about this?
> On the passing side of an OSR guard, one or more fear points will be reachable without going through another OSR guard. On failure, the guard will transfer control to an OSR induce block.

gita-omr replied:
Done.

gita-omr commented on doc/compiler/osr/OSR.md:
fixed

gita-omr commented on doc/compiler/osr/OSR.md:
done

gita-omr commented on doc/compiler/osr/OSR.md:
done

gita-omr commented on doc/compiler/osr/OSR.md:
used your wording.

gita-omr replied:
applied

gita-omr commented:
I think addressed all comments (with some questions). Please take a look.

gita-omr commented:
Fixed line endings.

gita-omr commented:
I think I addressed all the comments, but some discussion might be needed for these:

[1] https://github.com/eclipse-openj9/openj9/pull/18596#discussion_r1439897093
[2] https://github.com/eclipse-openj9/openj9/pull/18596#discussion_r1439900372
[3] https://github.com/eclipse-openj9/openj9/pull/18596#discussion_r1440050989
[4] https://github.com/eclipse-openj9/openj9/pull/18596#discussion_r1440038800

jdmpapin commented on doc/compiler/hcr/NextGenHCR.md:
See *the OpenJ9* OSR document

jdmpapin commented on doc/compiler/osr/OSR.md:
```diff
@@ -0,0 +1,275 @@
+
+# Introduction
+
+This document describes several OpenJ9 optimizations that use the OSR mechanism described in [OMR OSR document](https://github.com/eclipse/omr/blob/master/doc/compiler/osr/OSR.md). These optimizations include:
```
described in *the* OMR OSR document

jdmpapin commented on doc/compiler/osr/OSR.md:
> that can be violated at an OSR yield point

I'd like to distinguish between invalidating an assumption and violating it. First we invalidate the assumption (i.e. stop assuming that it holds), and only after that can we violate it (i.e. change state so that it no longer holds)

If we make this distinction, then we should say that assumptions can be *invalidated* at an OSR yield point

jdmpapin commented on doc/compiler/osr/OSR.md:
the OSR induce block ~allow~ *allows* bookkeeping

jdmpapin commented on doc/compiler/osr/OSR.md:
*optimizaitons* → *optimizations*

(though I've suggested a change that would avoid this word here if you choose to use it or something similar)

jdmpapin commented on doc/compiler/osr/OSR.md:
*The* OSR induce block will contain *a* call to...

jdmpapin replied:
Still need to add *a*: will contain *a* call to

gita-omr replied:
done

jdmpapin commented on doc/compiler/osr/OSR.md:
I believe this exception edge to catch 1 should be removed as well

gita-omr commented:
Thanks a lot for taking another look. I think I addressed all the latest comments.

jdmpapin commented:
Thanks! Still missing an article from one of the little copy edits (https://github.com/eclipse-openj9/openj9/pull/18596#discussion_r1446386095), and I think you might not have seen https://github.com/eclipse-openj9/openj9/pull/18596#discussion_r1445295070. No idea why GitHub is so keen to hide review comments :roll_eyes:

gita-omr commented:
Pushed by mistake...

gita-omr commented:
Now pushed two latest corrections.

hzongaro commented on doc/compiler/osr/OSR.md:
I think jitInduceOSRAtCurrentPC should appear in bold type.

hzongaro commented on doc/compiler/osr/OSR.md:
I think BCI should be written out as "bytecode indices (BCIs)" here.  It might be good to add a link to a definition, if we have one someplace.

gita-omr replied:
Interestingly, could not find a definition in the openj9 docs.

jdmpapin replied:
BCI here means the `TR_ByteCodeInfo` rather than bytecode index. It's really too bad that they abbreviate the same way to create this ambiguity

gita-omr replied:
Right, but I think the intent is clear  - we associate some info with BCI.

hzongaro commented on doc/compiler/osr/OSR.md:
Maybe instead of "When OSR on guard failure is enabled" this should say "If OSR on guard failure is enabled. . . . "

hzongaro commented on doc/compiler/osr/OSR.md:
I would suggest making the CFGs code blocks and having the introductory lines, like "After ILGen:" in ordinary type.

hzongaro reviewed (commented):
Thanks for all this work on enhancing the documentation of OSR!

One thing that I noticed is that references to helpers, options, environment variables, flags in the compiler, etc. appear in bold type or in regular type.  I think they should be formatted as code.  I haven't gone through and commented on each of those, but may I ask you to try to clean that up.

gita-omr commented:
Pushed minor corrections to the latest commit.

gita-omr commented:
Fixed a link.

gita-omr commented:
Addressed all the latest comments from @hzongaro .

gita-omr commented:
Changed all method and class names to use code font.

gita-omr commented:
More font corrections.

gita-omr commented:
Some more formatting.

gita-omr commented:
Fixed line ending.

gita-omr commented:
Fixed line ending.

gita-omr commented:
Introduced BCI earlier.

gita-omr commented:
Added `osrFearPointHelper` description under "Supporting Classes and Helpers" since it was in the original OMR document.

gita-omr commented:
Minor improvement.

hzongaro commented on doc/compiler/osr/OSR.md:
I think this link is broken.  I think it should be `../optimizer/LocalOptimizationsSummary.md#Rematerialization` - that is, it should only go up to the parent `doc/compiler` rather than up to the grandparent `doc`

gita-omr commented:
Addressed latest comments from @hzongaro

gita-omr commented:
Created a placeholder MethodHandles.md and pointed to it from README.md

hzongaro reviewed (approved):
Looks good!  Once you've squashed the commits, I will merge.

gita-omr commented:
Squashed.
